device.go

```diff
@@ -0,0 +1,267 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+package virtcontainers
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+
+	"github.com/go-ini/ini"
+)
+
+// Defining this as a variable instead of a const, to allow
+// overriding this in the tests.
+var sysIOMMUPath = "/sys/kernel/iommu_groups"
+
+var sysDevPrefix = "/sys/dev"
+
+var blockPaths = []string{"/dev/sd", "/dev/hd", "/dev/vd", "/dev/ida/"}
+
+const (
+	vfioPath = "/dev/vfio/"
+)
+
+// Device is the virtcontainers device interface.
+type Device interface {
+	attach(hypervisor) error
+	detach(hypervisor) error
+}
+
+// DeviceInfo is an embedded type that contains device data common to all types of devices.
+type DeviceInfo struct {
+	// Device path on host
+	HostPath string
+
+	// Device path inside the container
+	ContainerPath string
+
+	// Type of device: c, b, u or p
+	// c , u - character(unbuffered)
+	// p - FIFO
+	// b - block(buffered) special file
+	// More info in mknod(1).
+	DevType string
+
+	// Major, minor numbers for device.
+	Major int64
+	Minor int64
+
+	// FileMode permission bits for the device.
+	FileMode *os.FileMode
+
+	// id of the device owner.
+	UID uint32
+
+	// id of the device group.
+	GID uint32
+}
+
+// VFIODevice is a vfio device meant to be passed to the hypervisor
+// to be used by the Virtual Machine.
+type VFIODevice struct {
+	DeviceInfo
+	BDF string
+}
+
+func newVFIODevice(devInfo DeviceInfo) *VFIODevice {
+	d := &VFIODevice{}
+	d.DeviceInfo = devInfo
+	return d
+}
+
+func (device *VFIODevice) attach(h hypervisor) error {
+	vfioGroup := filepath.Base(device.HostPath)
+	iommuDevicesPath := filepath.Join(sysIOMMUPath, vfioGroup, "devices")
+
+	deviceFiles, err := ioutil.ReadDir(iommuDevicesPath)
+	if err != nil {
+		return err
+	}
+
+	// Pass all devices in iommu group
+	for _, deviceFile := range deviceFiles {
+
+		//Get bdf of device eg 0000:00:1c.0
+		deviceBDF, err := bdf(deviceFile.Name())
+		if err != nil {
+			return err
+		}
+
+		device.BDF = deviceBDF
+
+		if err := h.addDevice(device, vfioDev); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (device *VFIODevice) detach(h hypervisor) error {
+	return nil
+}
+
+// BlockDevice refers to a block storage device implementation.
+type BlockDevice struct {
+	DeviceInfo
+}
+
+func newBlockDevice(devInfo DeviceInfo) *BlockDevice {
+	d := &BlockDevice{}
+	d.DeviceInfo = devInfo
+	return d
+}
+
+func (device *BlockDevice) attach(h hypervisor) error {
```

Comment:
Why is this not implemented ? My idea would be that `addDrive()` should be called from there, right ? I mean we are trying to manage the different type of devices through an interface, meaning that we want to remove the specific calls from `container.start()` dedicated to block devices, right ?

Reply:
I mentioned in the PR that I will be handling just vfio devices here. Handling block devices is a separate feature and should be done separately. It involves quite a few cases to handle and test.